### DISCUSSION TITLE
Invoke action at certain time

### DISCUSSION THREAD

**Discussion by fakhrulhilal** [jamesmh/coravel#307] [Q&A]
I want to execute certain action at given time with payload. I don't know what is the best way to do this, either using queue or task scheduler. Example, remove blob file at path 'xxx' at 2020-09-05T20:05:10Z0. So the payload will be blob path.

**Comment by jamesmh** [comment]
Use the task scheduler. For a specific date, you'll want to probably use the `Cron()` function for scheduling.

Use the typical cron format like `05 20 05 09 *` -> that would run you job on Sept. 5th at hour 20 minute 5 in UTC timezone (like in your example).

Now, it depends where the payload is coming from. If these are dynamically / user generate payloads, then you might want to do something more like this:

1. Use the queue to fire off the first piece of work (eg. storing the blob file).
2. Generate and store a mapping of `blob + expiry-date` (in a database most likely).
3. Use the scheduler to check every X minutes or every hour, or every day. Run a query against the database to see if there are any expired blobs - and expire them if there are.

**Reply by fakhrulhilal** [reply]
> 

AFAIK, cron syntax only support until "minute" unit, unless we use [NCrontab](https://github.com/atifaziz/NCrontab) which supports "second" unit. Am I correct that the example you provide will not run every year? My understanding, the job should be executed once (fire and forget), like a [delayed job in hangfire](https://docs.hangfire.io/en/latest/background-methods/calling-methods-with-delay.html). The payload should be dynamic. The caller will send blob path which (probably) queried from db.

**Comment by jamesmh** [comment]
The cron expression I gave as an example would run every year. If you stored the expiration date as in the second suggestion, then the scheduler would run more often (every minute or second if you wanted), but the item in question wouldn't be removed until the expiration date is reached.

**Comment by fakhrulhilal** [comment]
To be precise, I am looking for something like fire and forget, that could be `scheduler.ScheduleOnce(DateTime)` or `scheduler.Next(TimeSpan)`. It's likely can use workaround 
```cs
var expirationTime = DateTime.UtcNow.AddDay(1); // or find it from db
var scheduleTime = expirationTime - DateTime.UtcNow;
queue.QueueAsyncTask(() => {
    await Task.Delay(scheduleTime);
    // remove the file
});

**Comment by jamesmh** [comment]
Okay - this functionality doesn't exist yet. Would require persistent queued item which I'd like to rig up 👍
